crates/sui-framework/packages/move-stdlib/sources/vector.move

```diff
@@ -158,4 +158,149 @@ module std::vector {
         v.swap(i, last_idx);
         v.pop_back()
     }
+
+    // === Macros ===
+
+    /// Destroy the vector `v` by calling `f` on each element and then destroying the vector.
+    /// Does not preserve the order of elements in the vector (starts from the end of the vector).
+    public macro fun destroy<$T>($v: vector<$T>, $f: |$T|) {
+        let mut v = $v;
+        while (!v.is_empty()) {
+            $f(v.pop_back());
+        };
+        v.destroy_empty();
+    }
+
+    /// Destroy the vector `v` by calling `f` on each element and then destroying the vector.
+    /// Preserve the order of elements in the vector.
+    public macro fun do<$T>($v: vector<$T>, $f: |$T|) {
+        let mut v = $v;
+        v.reverse();
+        while (!v.is_empty()) {
+            $f(v.pop_back());
+        };
+        v.destroy_empty();
+    }
+
+    /// Perform an action `f` on each element of the vector `v`. The vector is not modified.
+    public macro fun do_ref<$T>($v: &vector<$T>, $f: |&$T|) {
+        let v = $v;
+        let mut i = 0;
+        let len = v.length();
+        while (i < len) {
+            $f(&v[i]);
+            i = i + 1;
+        };
```

Comment:
```suggestion
        v.length().do!(|i| $f(&v[i]))     
```

Reply:
Now I see what you mean. Yeah, `do!` makes a lot of sense.